File: commit-msg.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Main execution
if ARGV.empty?
  puts "Error: No commit message file provided."
  exit(1)
end

SUBJECT_MAX_LENGTH = 50
BODY_MAX_LENGTH = 72
ERRORS  = []

def clean_line(line)
  line.start_with?("#") ? nil : line
end

def read_commit_message(file_path)
  File.readlines(file_path, chomp: true).map { |line| clean_line(line.strip) } .compact
end

def validate_line_length(line, length = 0)
  !line.nil? && line.length <= length
end

def validate_commit_message(message)

  if message[0].nil? || message[0].empty?
    ERRORS << "Commit message is missing a Subject line"
  end

  if !message[0].downcase.start_with?('merge') && !validate_line_length(message[0], SUBJECT_MAX_LENGTH)
    ERRORS << "Subject must be less than #{SUBJECT_MAX_LENGTH} characters"
  end

  if !message[1].nil? && !message[1].empty?
    ERRORS << "There should be a blank line after Subject"
  end

  unless message[2].nil?
    valid_body = message[2..].reject(&:empty?).map { |l| validate_line_length l, BODY_MAX_LENGTH }.all?
    ERRORS << "Some lines in the body of your message exceed the limit of #{BODY_MAX_LENGTH} characters" unless valid_body
  end
end

# Commit message verification
commit_message = read_commit_message(ARGV[0])
validate_commit_message commit_message

if commit_message.reject(&:empty?).empty?
  puts "Commit message cannot be left blank."
  exit(1)
end

unless ERRORS.empty?
  puts 'There are errors in your commit message:'
  puts ERRORS.join("\n")
  exit(1)
end

File: commit-msg.py
#!/usr/bin/env python
# coding: utf-8

import platform
import sys


if len(sys.argv) < 2:
    print(f"Error: No commit message file provided.\nArgs({len(sys.argv)}): {sys.argv}")
    exit(1)


SUBJECT_MAX_LENGTH = 50
BODY_MAX_LENGTH = 72
ERRORS  = []


def clean_line(line):
    line = line.strip()
    if len(line):
        if line[0] == "#":
            return None
        if line[-1] == "\n":
            return line[:-1]
    return line


def in_venv():
    return sys.prefix != sys.base_prefix


def read_commit_msg():
    with open(sys.argv[1]) as f:
        return [clean_line(l) for l in f.readlines()]


def validate_line_length(line, length=0):
    return len(line) <= length


def validate_commit_message(message):
    try:
        if not message[0]:
            ERRORS.append("Commit message is missing a Subject line")
        elif not message[0].lower().startswith('merge') and not validate_line_length(message[0], SUBJECT_MAX_LENGTH):
            ERRORS.append(f"Subject must be less than {SUBJECT_MAX_LENGTH} characters")

        if message[1]:
            ERRORS.append("There should be a blank line after Subject")

        if message[2]:
            valid_body = all([validate_line_length(item, BODY_MAX_LENGTH) for item in message[2:] if item not in (None, '')])
            ERRORS.append(f"Some lines in the body of your message exceed the limit of {BODY_MAX_LENGTH} characters") if not valid_body else None

    except (IndexError, TypeError):
        pass

# Commit message verification
print(f"Your python version is {platform.python_version()} ({ '' if in_venv() else 'not ' }running in virtualenv).")

validate_commit_message(read_commit_msg())

if len(ERRORS):
    print('There are errors in your commit message:')
    print('\n'.join(ERRORS))
    exit(1)
